File: brixen-dotnet/test/bean/ContentContainerBeanTest.cs
using System;
using NUnit.Framework;
using Org.Brixen.Bean;
using OpenQA.Selenium;
using Moq;

namespace Org.Brixen.Bean.Tests {

	[TestFixture]
	[Category("ContentcContainerBeanTestGroup")]
	public class ContentContainerBeanTest {

		[Test]
		[Category("ExpectedExceptionsTestGroup")]
		[Category("BeanExpectedExceptionsTestGroup")]
		[Category("ContentContainerBeanExpectedExceptionsTestGroup")]
		[ExpectedException("System.ArgumentNullException", ExpectedMessage = "Cannot invoke setter for " + 
			"ContentContainer property with a null parameter", MatchType=MessageMatch.Regex)]
		public void shouldThrowExceptionForNullContainerElement() {
			IContentContainerBean bean = new ContentContainerBean();
			bean.ContentContainer = null;
		}

		[Test]
		[Category("EqualityNegativeTestGroup")]
		[Category("BeanEqualityNegativeTestGroup")]
		[Category("ContentContainerBeanEqualityNegativeTestGroup")]
		public void ShouldNotBeEqual() {
			Mock<IWebDriver> mockDriverOne = new Mock<IWebDriver>();
			Mock<IWebDriver> mockDriverTwo = new Mock<IWebDriver>();

			IContentContainerBean beanOne = new ContentContainerBean();
			IContentContainerBean  beanTwo = new ContentContainerBean();

			beanOne.Driver = mockDriverOne.Object;
			beanTwo.Driver = mockDriverTwo.Object;

			Assert.AreNotEqual(beanOne, beanTwo, "Beans with different driver references should not be equal");

			beanOne = new ContentContainerBean();
			beanTwo = new ContentContainerBean();
			beanOne.LoadTimeout = 40;

			Assert.AreNotEqual(beanOne, beanTwo, "Beans with different load timeouts should not be equal");

			Mock<IWebElement> mockContainerElementOne = new Mock<IWebElement>();
			Mock<IWebElement> mockContainerElementTwo = new Mock<IWebElement>();

			beanOne.LoadTimeout = 30;
			beanOne.ContentContainer = mockContainerElementOne.Object;

			Assert.AreNotEqual(beanOne, beanTwo, "Beans with different content container element values should not " + 
				"be equal");

			beanTwo.ContentContainer = mockContainerElementTwo.Object;

			Assert.AreNotEqual(beanOne, beanTwo, "Beans with different content container element values should not " + 
				"be equal");
		}

		[Test]
		[Category("EqualityPositiveTestGroup")]
		[Category("BeanEqualityPositiveTestGroup")]
		[Category("ContentContainerBeanEqualityPositiveTestGroup")]
		public void ShouldBeEqual() {
			IContentContainerBean beanOne = new ContentContainerBean();
			IContentContainerBean beanTwo = new ContentContainerBean();

			Assert.AreEqual(beanOne, beanTwo, "Two newly constructed beans should be equal before any setters are " + 
				"invoked");

			Mock<IWebDriver> mockDriver = new Mock<IWebDriver>();

			beanOne.Driver = mockDriver.Object;
			beanTwo.Driver = mockDriver.Object;

			Assert.AreEqual(beanOne, beanTwo, "Two beans should be equal if they have the same web driver reference " +
				"and the default load timeout");

			Mock<IWebElement> mockContainerElement = new Mock<IWebElement>();

			beanOne.ContentContainer = mockContainerElement.Object;
			beanTwo.ContentContainer = mockContainerElement.Object;

			Assert.AreEqual(beanOne, beanTwo, "Two beans should be equal if they have the same web driver and " + 
				"content container references");

			beanOne = new ContentContainerBean();
			beanTwo = new ContentContainerBean();

			beanOne.ContentContainer = mockContainerElement.Object;
			beanTwo.ContentContainer = mockContainerElement.Object;

			Assert.AreEqual(beanOne, beanTwo, "Two beans should be equal if they have null web driver references " +
				"and the same content container reference");
		}

		[Test]
		[Category("ToStringCallsBaseTestGroup")]
		[Category("BeanToStringCallsBaseTestGroup")]
		[Category("ContentContainerBeanToStringCallsBaseTestGroup")]
		public void ShouldCallBaseForToString() {
			Mock<IWebDriver> mockDriver = new Mock<IWebDriver>();
			Mock<IWebElement> mockContainerElement = new Mock<IWebElement>();

			IContentContainerBean bean = new ContentContainerBean();
			bean.Driver = mockDriver.Object;
			bean.ContentContainer = mockContainerElement.Object;

			Assert.AreEqual(bean.ToString(), "ContentContainerBean(LoadableBean(Driver: " + 
				mockDriver.Object.ToString() + ", LoadTimeout: 30), " + "ContentContainer: " + 
				mockContainerElement.Object.ToString() + ")");
		}

		[Test]
		[Category("HashCodeCallsSuperTestGroup")]
		[Category("BeanHashCodeCallsSuperTestGroup")]
		[Category("ContentContainerBeanHashCodeCallsSuperTestGroup")]
		public void ShouldCallBaseForHashCode() {
			IContentContainerBean bean = new ContentContainerBean();
			IContentContainerBean beanToCompare = new ContentContainerBean();

			Assert.AreEqual(bean.GetHashCode(), beanToCompare.GetHashCode(), "Hash codes for bean which have not had " + 
				"setters called should be equal, but are not: " + bean.ToString() + ", " + beanToCompare.ToString());

			Mock<IWebDriver> mockDriver = new Mock<IWebDriver>();

			bean.Driver = mockDriver.Object;
			Assert.AreNotEqual(bean.GetHashCode(), beanToCompare.GetHashCode(), "Hash codes for bean which have " + 
				"different values for their driver fields should not be equal, but are: " + bean.ToString() + ", " +
				beanToCompare.ToString());

			beanToCompare.Driver = mockDriver.Object;
			Assert.AreEqual(bean.GetHashCode(), beanToCompare.GetHashCode(), "Hash codes for bean which have the " + 
				"same driver should be equal, but are not: " + bean.ToString() + ", " + beanToCompare.ToString());

			beanToCompare.LoadTimeout = 200;
			Assert.AreNotEqual(bean.GetHashCode(), beanToCompare.GetHashCode(), "Hash codes for bean which have " + 
				"different values for their load timeout fields should not be equal, but are: " + bean.ToString() + 
				", " + beanToCompare.ToString());

			bean.LoadTimeout = 200;
			Assert.AreEqual(bean.GetHashCode(), beanToCompare.GetHashCode(), "Hash codes for bean which have the " + 
				"same driver and load timeout should be equal, but are not: " + bean.ToString() + ", " + 
				beanToCompare.ToString());
		}
	}
}

File: brixen-java/src/test/java/org/brixen/test/bean/ContentContainerBeanTest.java
package org.brixen.test.bean;

import org.brixen.bean.ContentContainerBean;
import org.brixen.bean.ContentContainerBeanImpl;
import org.mockito.Mock;
import org.mockito.MockitoAnnotations;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.WebElement;
import org.testng.annotations.BeforeMethod;
import org.testng.annotations.Test;

import static org.mockito.Mockito.when;
import static org.testng.Assert.assertEquals;
import static org.testng.Assert.assertNotEquals;
import static org.testng.Assert.assertTrue;

@Test(groups = {"LoadableBeanTestGroup"})
public class ContentContainerBeanTest {
    private @Mock WebDriver mockDriverOne;
    private @Mock WebDriver mockDriverTwo;

    private @Mock WebElement mockElementOne;
    private @Mock WebElement mockElementTwo;

    @BeforeMethod
    public void setUp() {
        MockitoAnnotations.initMocks(this);
    }

    @SuppressWarnings("ConstantConditions")
    @Test(groups = {"ExpectedExceptionsTestGroup", "BeanExpectedExceptionsTestGroup",
            "ContentContainerBeanExpectedExceptionsTestGroup"},
            expectedExceptions = { NullPointerException.class }
    )
    public void shouldThrowExceptionForNullContainerElement() {
        final ContentContainerBean bean = new ContentContainerBeanImpl();
        bean.setContentContainer(null);
    }

    @Test(groups = {"EqualityNegativeTestGroup", "BeanEqualityNegativeTestGroup",
            "ContentContainerBeanEqualityNegativeTestGroup"}
    )
    public void shouldNotBeEqual() {
        ContentContainerBean beanOne = new ContentContainerBeanImpl();
        ContentContainerBean  beanTwo = new ContentContainerBeanImpl();

        beanOne.setDriver(mockDriverOne);
        beanTwo.setDriver(mockDriverTwo);

        assertNotEquals(beanOne, beanTwo, "Beans with different driver references should not be equal");

        beanOne = new ContentContainerBeanImpl();
        beanTwo = new ContentContainerBeanImpl();
        beanOne.setLoadTimeout(40);

        assertNotEquals(beanOne, beanTwo, "Beans with different load timeouts should not be equal");

        beanOne.setLoadTimeout(30);
        beanOne.setContentContainer(mockElementOne);

        assertNotEquals(beanOne, beanTwo, "Beans with different content container element values should not be equal");

        beanTwo.setContentContainer(mockElementTwo);

        assertNotEquals(beanOne, beanTwo, "Beans with different content container element values should not be equal");
    }

    @Test(groups = {"EqualityPositiveTestGroup", "BeanEqualityPositiveTestGroup",
            "ContentContainerBeanEqualityPositiveTestGroup"}
    )
    public void shouldBeEqual() {
        ContentContainerBean beanOne = new ContentContainerBeanImpl();
        ContentContainerBean beanTwo = new ContentContainerBeanImpl();

        assertEquals(beanOne, beanTwo, "Two newly constructed beans should be equal before any setters are invoked");

        beanOne.setDriver(mockDriverOne);
        beanTwo.setDriver(mockDriverOne);

        assertEquals(beanOne, beanTwo, "Two beans should be equal if they have the same web driver reference and the " +
                "default load timeout");

        beanOne.setContentContainer(mockElementOne);
        beanTwo.setContentContainer(mockElementOne);

        assertEquals(beanOne, beanTwo, "Two beans should be equal if they have the same web driver and content " +
                "container references");

        beanOne = new ContentContainerBeanImpl();
        beanTwo = new ContentContainerBeanImpl();

        beanOne.setContentContainer(mockElementOne);
        beanTwo.setContentContainer(mockElementOne);

        assertEquals(beanOne, beanTwo, "Two beans should be equal if they have null web driver references and the " +
                "same content container reference");
    }

    @Test(groups = {"ToStringCallsSuperTestGroup", "BeanToStringCallsSuperTestGroup",
            "ContentContainerBeanToStringCallsSuperTestGroup"}
    )
    public void shouldCallSuperForToString() {
        final ContentContainerBean bean = new ContentContainerBeanImpl();

        when(mockDriverOne.toString()).thenReturn("Mock WebDriver");
        bean.setDriver(mockDriverOne);
        bean.setLoadTimeout(100);

        assertTrue(bean.toString().contains("driver=Mock WebDriver"), "toString() for ContentContainerBeanImpl does " +
                "not include the driver field in LoadableBeanImpl: " + bean.toString());
        assertTrue(bean.toString().contains("loadTimeout=100"), "toString() for ContentContainerBeanImpl does " +
                "not include the loadTimeoutInSeconds field in LoadableBeanImpl: " + bean.toString());
    }

    @Test(groups = {"HashCodeCallsSuperTestGroup", "BeanHashCodeCallsSuperTestGroup",
            "ContentContainerBeanHashCodeCallsSuperTestGroup"},
            dependsOnMethods = {"shouldCallSuperForToString"}
    )
    public void shouldCallSuperForHashCode() {
        final ContentContainerBean bean = new ContentContainerBeanImpl();
        final ContentContainerBean beanToCompare = new ContentContainerBeanImpl();

        assertEquals(bean.hashCode(), beanToCompare.hashCode(), "Hash codes for bean which have not had setters " +
                "called should be equal, but are not: " + bean.toString() + ", " + beanToCompare.toString());

        bean.setDriver(mockDriverOne);
        assertNotEquals(bean.hashCode(), beanToCompare.hashCode(), "Hash codes for bean which have different values " +
                "for their driver fields should not be equal, but are: " + bean.toString() + ", " +
                beanToCompare.toString());

        beanToCompare.setDriver(mockDriverOne);
        assertEquals(bean.hashCode(), beanToCompare.hashCode(), "Hash codes for bean which have the same driver " +
                "should be equal, but are not: " + bean.toString() + ", " + beanToCompare.toString());

        beanToCompare.setLoadTimeout(200);
        assertNotEquals(bean.hashCode(), beanToCompare.hashCode(), "Hash codes for bean which have different values " +
                "for their load timeout fields should not be equal, but are: " + bean.toString() + ", " +
                beanToCompare.toString());

        bean.setLoadTimeout(200);
        assertEquals(bean.hashCode(), beanToCompare.hashCode(), "Hash codes for bean which have the same driver and " +
                "load timeout should be equal, but are not: " + bean.toString() + ", " + beanToCompare.toString());
    }

    @Test(groups = {"EqualsCallsSuperTestGroup", "BeanEqualsCallsSuperTestGroup",
            "ContentContainerBeanEqualsCallsSuperTestGroup"},
            dependsOnMethods = {"shouldCallSuperForToString", "shouldCallSuperForHashCode"}
    )
    public void shouldCallSuperForEquals() {
        final ContentContainerBean bean = new ContentContainerBeanImpl();
        final ContentContainerBean beanToCompare = new ContentContainerBeanImpl();

        assertEquals(bean, beanToCompare, "Equals method should return true, but did not: " + bean.toString() + ", " +
                beanToCompare.toString());

        bean.setDriver(mockDriverOne);
        assertNotEquals(bean, beanToCompare, "Equals method should return false, but did not: " + bean.toString() +
                ", " + beanToCompare.toString());

        beanToCompare.setDriver(mockDriverOne);
        assertEquals(bean, beanToCompare, "Equals method should return true, but did not: " + bean.toString() + ", " +
                beanToCompare.toString());

        bean.setLoadTimeout(100);
        assertNotEquals(bean, beanToCompare, "Equals method should return false, but did not: " + bean.toString() +
                ", " + beanToCompare.toString());

        beanToCompare.setLoadTimeout(100);
        assertEquals(bean, beanToCompare, "Equals method should return true, but did not: " + bean.toString() + ", " +
                beanToCompare.toString());
    }
}
